packages/@eventual/aws-cdk/src/bucket-service.ts

```diff
@@ -0,0 +1,247 @@
+          serviceProps: props,
+        }),
+      ])
+    ) as ServiceBuckets<Service>;
+
+    this.bucketStreams = Object.values(
+      this.buckets as Record<string, Bucket>
+    ).reduce((streams: Record<string, BucketStream>, ent) => {
+      return {
+        ...streams,
+        ...ent.streams,
+      };
+    }, {}) as ServiceBucketStreams<Service>;
+  }
+
+  public configureReadWriteBuckets(func: Function) {
+    this.addEnvs(func, ENV_NAMES.SERVICE_NAME);
+    this.grantReadWriteBuckets(func);
+  }
+
+  public grantReadWriteBuckets(grantee: IGrantable) {
+    // grants the permission to start any task
+    grantee.grantPrincipal.addToPrincipalPolicy(
+      new PolicyStatement({
+        actions: ["s3:List*"],
+        resources: [
+          serviceBucketArn(
+            this.props.serviceName,
+            bucketServiceBucketSuffix("*"),
+            false
+          ),
+        ],
+      })
+    );
+
+    grantee.grantPrincipal.addToPrincipalPolicy(
+      new PolicyStatement({
+        actions: [
+          "s3:List*",
+          "s3:GetObject*",
+          "s3:PutObject*",
+          "s3:DeleteObject*",
+        ],
+        resources: [
+          `${serviceBucketArn(
+            this.props.serviceName,
+            bucketServiceBucketSuffix("*"),
+            false
+          )}/*`,
+        ],
+      })
+    );
+  }
+
+  private readonly ENV_MAPPINGS = {
+    [ENV_NAMES.SERVICE_NAME]: () => this.props.serviceName,
+  } as const;
+
+  private addEnvs(func: Function, ...envs: (keyof typeof this.ENV_MAPPINGS)[]) {
+    envs.forEach((env) => func.addEnvironment(env, this.ENV_MAPPINGS[env]()));
+  }
+}
+
+interface BucketProps {
+  serviceProps: BucketServiceProps<any>;
+  bucketService: BucketService<any>;
+  bucket: BucketRuntime;
+}
+
+export class Bucket extends Construct {
+  public bucket: s3.Bucket;
+  public streams: Record<string, BucketStream>;
+
+  constructor(scope: Construct, props: BucketProps) {
+    super(scope, props.bucket.name);
+    console.log(s3);
+    this.bucket = new s3.Bucket(this, "Bucket", {
+      cors:
+        props.serviceProps.cors &&
+        props.serviceProps.cors.allowMethods &&
+        props.serviceProps.cors.allowOrigins
+          ? [
+              {
+                allowedMethods: props.serviceProps.cors.allowMethods.flatMap(
+                  (s) =>
+                    Object.values(s3.HttpMethods).find(
+                      (v) => s.toString() === v.toString()
+                    ) ?? []
+                ),
+                allowedOrigins: props.serviceProps.cors.allowOrigins,
+                allowedHeaders: props.serviceProps.cors.allowHeaders,
+                exposedHeaders: props.serviceProps.cors.exposeHeaders,
+                maxAge: props.serviceProps.cors.maxAge?.toSeconds(),
+              },
+            ]
+          : undefined,
+      bucketName: bucketServiceBucketName(
+        props.serviceProps.serviceName,
+        props.bucket.name
+      ),
```

Comment:
bucket names are the most problematic because of global uniqueness. Do you have any thoughts on that?

Reply:
Just that the service name would make it more unique

Another idea was to use a single bucket and split internally, but that could create issues if trying to make a static site or cdn.

Or add a unique value to the name, which is already limited length.

Reply:
Perhaps we just allow the name to be overridden. Problem solved.